src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,656 @@
+        {
+            var nameToParameterDict = new Dictionary<TIdentifierNameSyntax, IParameterSymbol>();
+            var variablesInExpression = expression.DescendantNodes().OfType<TIdentifierNameSyntax>();
+            var semanticModel = document.SemanticModel;
+
+            foreach (var variable in variablesInExpression)
+            {
+                var symbol = semanticModel.GetSymbolInfo(variable, cancellationToken).Symbol;
+                if (symbol != null && symbol is IParameterSymbol parameterSymbol)
+                {
+                    if (!nameToParameterDict.ContainsKey(variable))
+                    {
+                        nameToParameterDict.Add(variable, parameterSymbol);
+                    }
+                }
+            }
+            return nameToParameterDict;
+        }
+
+        /// <summary>
+        /// Determines if the expression is contained within something that is "parameterized"
+        /// </summary>
+        private (bool isParameterized, bool hasOptionalParameter) ExpressionWithinParameterizedMethod(SemanticDocument semanticDocument,
+                                                                                                      TExpressionSyntax expression,
+                                                                                                      CancellationToken cancellationToken)
+        {
+            var methodSymbol = GetMethodSymbolFromExpression(semanticDocument, expression, cancellationToken);
+
+            var variablesInExpression = expression.DescendantNodes().OfType<TIdentifierNameSyntax>();
+            var semanticModel = semanticDocument.SemanticModel;
+            var hasOptionalParameter = false;
+
+            foreach (var variable in variablesInExpression)
+            {
+                var parameterSymbol = semanticModel.GetSymbolInfo(variable, cancellationToken).Symbol;
+                if (parameterSymbol == null || parameterSymbol is not IParameterSymbol)
+                {
+                    return (false, hasOptionalParameter);
+                }
+                if (((IParameterSymbol)parameterSymbol).IsParams)
+                {
+                    return (false, hasOptionalParameter);
+                }
+                if (((IParameterSymbol)parameterSymbol).HasExplicitDefaultValue)
+                {
+                    hasOptionalParameter = true;
+                }
+            }
+            return (methodSymbol != null && methodSymbol.GetParameters().Any(), hasOptionalParameter);
+        }
+
+        private (string title, ImmutableArray<CodeAction> actions) AddActions(SemanticDocument semanticDocument, TExpressionSyntax expression, CancellationToken cancellationToken)
+        {
+            var actionsBuilder = new ArrayBuilder<CodeAction>();
+            var (isParameterized, hasOptionalParameter) = ExpressionWithinParameterizedMethod(semanticDocument, expression, cancellationToken);
+            if (isParameterized)
```

Comment:
i don't understand this.  if it's not parameterized... why do anything?

Reply:
It returns an empty actionsBuilder, which if it is empty returns null where it is called